### ISSUE TITLE
Grin node 5.1.0 won't launch on Linux

### ISSUE THREAD

**Issue by nthrow** [mimblewimble/grin#3641]
**Describe the bug**
Grin node v5.1.0 will not launch on Debian Buster, returning "Illegal instruction" to stdout.

**To Reproduce**
```
$ grin
Illegal instruction
$ echo $?
132
$ ps aux |grep -E '[g]rin'
$ 
```

**Relevant Information**
```
20210506 16:53:04.446 INFO grin_util::logger - log4rs is initialized, file level: Debug, stdout level: Warn, min. level: Debug                                                                                      
20210506 16:53:04.447 INFO grin - Using configuration file at /home/$USER/.grin/main/grin-server.toml
20210506 16:53:04.447 INFO grin - This is Grin version 5.1.0 (git v5.1.0), built for x86_64-unknown-linux-gnu by rustc 1.51.0 (2fd73fabe 2021-03-23).                                                               
20210506 16:53:04.448 DEBUG grin - Built with profile "release", features "".
20210506 16:53:04.448 INFO grin - Chain: Mainnet
20210506 16:53:04.448 INFO grin - Accept Fee Base: 500000
20210506 16:53:04.449 INFO grin - Future Time Limit: 300
20210506 16:53:04.449 INFO grin - Feature: NRD kernel enabled: false
20210506 16:53:04.449 WARN grin::cmd::server - Starting GRIN in UI mode...
20210506 16:53:04.450 INFO grin_servers::grin::server - Starting server, genesis block: 40adad0aec27
20210506 16:53:04.451 DEBUG grin_store::lmdb - DB Mapsize for /home/$USER/.grin/main/chain_data/lmdb is 549755813888
```

**Desktop (please complete the following information):**
```
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

**Comment by tromp** [comment]
A failing assert while dealing with GlobalConfig... I suspect that's from here:

https://github.com/mimblewimble/grin/blob/9e27e6f9d3e960466b75f666057143c06e2f5cfc/config/src/config.rs#L345-L348

and related to the rewritten configuration grin-server.toml file.

**Comment by MCM-Mike** [comment]
>To Reproduce
>$ grin
>Illegal instruction
>$ echo $?
>132
>$ ps aux |grep -E '[g]rin'
>$ 

Did you GRIN setup the server using 

```
./grin server config
``` 
and started it as shown in `grin --help` ? 

```
_$ grin server -c grin-server.toml run 
```

**Comment by nthrow** [comment]
> 
> Did you setup the GRIN server using
> 
> ```
> ./grin server config
> ```

Yes, I have been using the same grin-server.toml essentially since the network launch.

**Comment by MCM-Mike** [comment]
>Yes, I have been using the same grin-server.toml essentially since the network launch.

Keep in mind there is a config change in `5.1.0` see https://github.com/mimblewimble/grin/pull/3634 where you need to change the config file slightly but this should not really be a problem, if I am correctly informed.

**Comment by nthrow** [comment]
Okay, prepending `config_file_version = 2` to the top of my grin-server.toml sorted the issue.  

As much as I love this project, this is extraordinarily bad user experience - I'm more than familiar with server configuration, code compilation, etc and yet something as simple as adding one line to the top of the config file isn't able to be gracefully handled by the node, or at least have proper error handling?

**Comment by MCM-Mike** [comment]
But to share my problems on `Debian` I had similar problems on Debian 9, but different error. 
As I am running quite a few nodes, it was essential that the nightly compiling where not interrupted and we could upgrade our main `grin-nodes` to latest version. Therefore I did some manual tests with Debian 9 which failed, I also was not changing anything nor upgrading any dependencies. 

On `Debian 9` when trying to compile like usually, I got this error:
`error: failed to run custom build command for croaring-sys v0.4.6` 
(see detail error log here: https://gist.github.com/MCM-Mike/e326734bed9e6965f95d2700179ec13c

What I ended up doing was upgrading all nodes to latest Debian 10 (buster) from Debian 9 using:
_(changing source list, etc this is not a tutorial on how to upgrade Debian)_
```
sudo apt update
sudo apt upgrade
sudo apt full-upgrade
sudo apt --purge autoremove
```

Then I updated rust using `/root/.cargo/bin/rustup update` and checked out the latest version from github via:

```
/usr/bin/git fetch
/usr/bin/git pull
/usr/bin/git checkout master
/usr/bin/git fetch --tags
latestTag=$(git tag --sort=refname | grep -v -e "beta" -e "rc"| tail -n 1)
/usr/bin/git checkout $latestTag
```

and compiled sucessfully on Debian 10 with:

```
/root/.cargo/bin/cargo clean
/root/.cargo/bin/cargo build --release
```

**current Debian 10:**
```cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

_$ uname -a
Linux XXXX.grinnode.live 4.19.0-16-amd64 #1 SMP Debian 4.19.181-1 (2021-03-19) x86_64 GNU/Linux
```

**Comment by antiochp** [comment]
Related: https://github.com/mimblewimble/grin/pull/3634

@trevyn Is there an edge case here where the config file migration is not behaving as expected?

**Comment by nthrow** [comment]
To clarify: adding the `config_file_version = 2` to the top of my grin-server.toml sorted the issue on Void Linux.  

Following up on Debian 10, adding that changed nothing.  Then, removing the old server toml, generating a new one via `grin server config`, adjusting what I needed to, and trying to start the server fails - but this time, with nothing returned to stderr or stdout, nor to the grin-server.log.

**Comment by phyro** [comment]
@nthrow grin-server.toml has two settings to control the level of logging: `stdout_log_level` and `file_log_level` which both default to values that don't tell everything they see. Perhaps changing it to "Debug" will give more information about the error. Make sure to check the debug doesn't leak any of your personal information if you decide to share it here.

**Comment by antiochp** [comment]
As far as I can tell the only thing that changed that _might_ cause something along the line of "Illegal Instruction" is -
https://github.com/mimblewimble/grin/pull/3596

@trevyn Any ideas if this may be related?

**Comment by bruges** [comment]
Same issue on Debian 11 (bullseye).
$ ./grin
Illegal instruction

I have "config_file_version = 2" line at the start of my grin-server.toml file.

**Comment by bladedoyle** [comment]
`20210518 09:09:47.542 ERROR grin_util::logger - 
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Chain(Error { inner: 

Other Error: output vs rproof MMRs different sizes })': src/bin/cmd/server.rs:57   0: grin_util::logger::send_panic_to_log::{{closure}}
`

**Comment by bladedoyle** [comment]
It looks like this is because we use the "-march=native" flag when building croaring.
A similar issue is described here:
https://github.com/mimblewimble/grin/issues/2494#issuecomment-521557508

```
bdoyle@lucy:~/dev/grin/target/release/build$ git status
HEAD detached at v5.1.0
bdoyle@lucy:~/dev/grin/target/release$ grep native ./build/croaring-sys-07b6100f886b91b1/output
running: "cc" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "-m64" "-Wall" "-Wextra" "-march=native" "-o" "/home/bdoyle/dev/grin/target/release/build/croaring-sys-07b6100f886b91b1/out/CRoaring/roaring.o" "-c" "CRoaring/roaring.c"
cargo:rustc-link-search=native=/home/bdoyle/dev/grin/target/release/build/croaring-sys-07b6100f886b91b1/out
```

**Comment by GeneFerneau** [comment]
I cannot reproduce this issue. On debian 10.8, I built `grin/v5.1.0`, created a new config, and started the node without issue.

What version of `clang` and `librust-clang-sys-dev` do you have installed?

**Comment by bladedoyle** [comment]
I dont believe this is related to linux version or rust version.  I think we are building an avx256 instruction into the binary (because croaring is built using native arch, and the CI/CD build system used to produce the grin release binary supports that instruction) which is not supported by very old CPU. 
I am able to reproduce it using an old laptop.

**Comment by GeneFerneau** [comment]
> I dont believe this is related to linux version or rust version. I think we are building an avx256 instruction into the binary (because croaring is built using native arch, and the CI/CD build system used to produce the grin release binary supports that instruction) which is not supported by very old CPU.

So, what CPU is @nthrow using? And do we want to bother supporting systems with such old CPUs?

**Comment by nthrow** [comment]
I'm not sure that's the issue as I'm currently on an AMD Ryzen 5 PRO 4650U.

**Comment by bruges** [comment]
My CPU is Intel i5-9400F

**Comment by quentinlesceller** [comment]
Can confirm the binary isn't working on my i7 running Debian 10.9 in docker. Trying to compile from source now.

**Comment by bladedoyle** [comment]
There may be more than one issue here?

I found that the first time running the new binary on an old cpu gives the "illegal instruction" crash.
Leading up to that crash, parts of the chain database have been written, but its left in a bad state.
Running grin again results in a different crash.
Its necessary to delete the bad chain database before being able to reproduce the illegal instruction crash.

```
root@crypto:~/.grin/main# rm -rf *
root@crypto:~/.grin/main# /tmp/grin-v5.1.0-official
Illegal instruction (core dumped)
root@crypto:~/.grin/main# /tmp/grin-v5.1.0-official
root@crypto:~/.grin/main# tail -n 40 grin-server.log
20210520 11:17:13.290 INFO grin_servers::grin::server - Starting server, genesis block: 40adad0aec27
20210520 11:17:13.290 INFO grin_store::lmdb - Resized database from 1048576 to 134217728
20210520 11:17:13.320 ERROR grin_util::logger - 
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Chain(Error { inner: 

Other Error: output vs rproof MMRs different sizes })': src/bin/cmd/server.rs:57   0: grin_util::logger::send_panic_to_log::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:595
   2: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:497
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:141
   4: rust_begin_unwind
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493
   5: core::panicking::panic_fmt
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:92
   6: core::option::expect_none_failed
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/option.rs:1300
   7: grin::cmd::server::start_server_tui
   8: grin::cmd::server::server_command
   9: grin::real_main
  10: grin::main
  11: std::sys_common::backtrace::__rust_begin_short_backtrace
  12: std::rt::lang_start::{{closure}}
  13: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ops/function.rs:259
      std::panicking::try::do_call
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:379
      std::panicking::try
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:343
      std::panic::catch_unwind
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panic.rs:431
      std::rt::lang_start_internal
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/rt.rs:51
  14: main
  15: __libc_start_main
  16: _start
root@crypto:~/.grin/main# rm -rf *
root@crypto:~/.grin/main# /tmp/grin-v5.1.0-official
Illegal instruction (core dumped)
```

**Comment by bladedoyle** [comment]
I think I read in the past that "march" isnt the only compiler flag that impacts which instructions are used.  The optimization setting (-On) can also cause AVX2 instructions to be used.

Here is a blurb that says something like that?
https://wiki.gentoo.org/wiki/GCC_optimization#-march
```
With -O2, the compiler will attempt to increase code performance without compromising on size, and without taking too much compilation time. SSE or AVX may be utilized at this level 
...
-O3 is also known to break several packages. Using -O3 is not recommended. However, it also enables -ftree-vectorize so that loops in the code get vectorized and will use AVX YMM registers.
```

Note that it appears croaring uses -O3 ?

-------

I also see that croaring (under the rs wrapper) supports a flag for not using AVX:
  https://github.com/RoaringBitmap/CRoaring/blob/master/CMakeLists.txt
 ```
option(ROARING_DISABLE_X64 "Forcefully disable x64 optimizations even if hardware supports it (this disables AVX)" OFF)
option(ROARING_DISABLE_AVX "Forcefully disable AVX even if hardware supports it " OFF)
(ROARING_DISABLE_NATIVE <span class="pl-s">"Forcefully disable -march optimizations (obsolete)"OFF")
```

We should try using those.

**Comment by tromp** [comment]
Thanks for investigating, Blade.

+1 on using ROARING_DISABLE_AVX

I guess that will add a -mno-avx flag to gcc.

**Comment by quentinlesceller** [comment]
Good catch @antiochp. This is now fixed by #3644. This binary should run fine on linux https://github.com/quentinlesceller/grin/releases/tag/v5.1.0-test3.
